sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/InternalObjectNode.java

```diff
@@ -103,17 +105,42 @@ public static Document fromObject(Object cosmosItem) {
         }
     }
 
-    public static ByteBuffer serializeJsonToByteBuffer(Object cosmosItem, ObjectMapper objectMapper) {
+    public static ByteBuffer serializeJsonToByteBuffer(Object cosmosItem, ObjectMapper objectMapper, String trackingId) {
         if (cosmosItem instanceof InternalObjectNode) {
-            return ((InternalObjectNode) cosmosItem).serializeJsonToByteBuffer();
+            InternalObjectNode internalObjectNode = ((InternalObjectNode) cosmosItem);
+            if (trackingId != null) {
+                internalObjectNode.set(Constants.Properties.TRACKING_ID, trackingId);
+            }
+            return internalObjectNode.serializeJsonToByteBuffer();
         } else if (cosmosItem instanceof Document) {
-            return ModelBridgeInternal.serializeJsonToByteBuffer((Document) cosmosItem);
+            Document doc = (Document) cosmosItem;
+            if (trackingId != null) {
+                doc.set(Constants.Properties.TRACKING_ID, trackingId);
+            }
+            return ModelBridgeInternal.serializeJsonToByteBuffer(doc);
         } else if (cosmosItem instanceof ObjectNode) {
-            return (new InternalObjectNode((ObjectNode)cosmosItem).serializeJsonToByteBuffer());
+            ObjectNode objectNode = (ObjectNode)cosmosItem;
+            if (trackingId != null) {
+                objectNode.put(Constants.Properties.TRACKING_ID, trackingId);
+            }
+            return (new InternalObjectNode(objectNode).serializeJsonToByteBuffer());
         } else if (cosmosItem instanceof byte[]) {
+            if (trackingId != null) {
+                InternalObjectNode internalObjectNode = new InternalObjectNode((byte[]) cosmosItem);
+                internalObjectNode.set(Constants.Properties.TRACKING_ID, trackingId);
+                return internalObjectNode.serializeJsonToByteBuffer();
+            }
             return ByteBuffer.wrap((byte[]) cosmosItem);
         } else {
-            return Utils.serializeJsonToByteBuffer(objectMapper, cosmosItem);
+            Object effectivePayload = cosmosItem;
+            if (trackingId != null) {
+                MapType mapType = objectMapper.getTypeFactory().constructMapType(LinkedHashMap.class,
```

Comment:
why are we converting to a map here?
Can we not convert cosmosItem into json string or byte[] and then create InternalObjectNode from it? 

InternalObjectNode internalObjectNode = new InternalObjectNode(objectMapper.writeValueAsBytes(cosmosItem)); 
or 
InternalObjectNode internalObjectNode = new InternalObjectNode(objectMapper.writeValueAsString(cosmosItem));

Reply:
We could - but that means less efficiency - serializing to json, deserializing and then serializing again.
This way we effectively deserializg into Map and then serialize after modification.